### PR TITLE
Fixes GitHub issue #30257

### PR DESCRIPTION
**Pull request by kirupha2000**
Fixes https://github.com/apache/beam/issues/30257

Adds a static `comparing` method to the `SerializableComparator` interface. This almost similar to the inbuilt Java's `Comparator#comparing(Function)` method except that it uses `SerializableFunction`.

### CODE REVIEW

**Comment by je-ik** [comment]
R: @je-ik

**Comment by kirupha2000** [comment]
Hi @je-ik , it took a while for me to figure out that there were some check style errors in the code😅. I've fixed those but looks like there are some checks failing that passed in the previous commit. Can you please guide me here if it's anything that I should fix on my end? Thanks!

**Comment by je-ik** [comment]
Hi @kirupha2000 , I'm sorry, I'm OOO, currently. I will take a look at this next week. Thanks for the contribution!

**Review by je-ik — COMMENTED** [review]
LGTM, I'd only suggest to test the serializable part of the contract of the created comparator.

**Comment by je-ik** [comment]
Also please squash the change to single commit. There is a documentation about how to structure granularity of commits in https://github.com/apache/beam/blob/master/contributor-docs/committer-guide.md. Thanks!

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SerializableComparatorTest.java
Comment (je-ik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SerializableComparatorTest.java:36]: This does not imply that the instance can be serialized. It would be better to use the `SerializableUtils` to check the serialization and deserialization.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java
Comment (je-ik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java:41]: I wonder, why `? super V` and not using `V` directly? Does it have any advantages?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java
Comment (je-ik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java:42]: Same here probably for `T`?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SerializableComparatorTest.java
Comment (je-ik) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SerializableComparatorTest.java:36]: You could have left the original tests as well, those were fine.

**Review by je-ik — COMMENTED** [review]
Added a few more comments.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java
Comment (kirupha2000) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java:41]: I think it does not have any advantages in this context since the `comparing` method here takes in a `SerializableFunction<? super T, ? extends V>`, meaning, because of the `? extends V` it would accept any type that is a sub type of `Comparable` as per my understanding.

But I had the `? super V`, following the "**P**roducers **E**xtends **C**onsumers **S**uper" general rule of thumb just in case to be safe.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java
Comment (kirupha2000) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java:42]: Same here, wasn't able to think of a scenario where this would be useful but added it anyways, following **PECS**, just in case.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SerializableComparatorTest.java
Comment (kirupha2000) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SerializableComparatorTest.java:36]: Would it be sufficient to check if the returned `Comparator` is serializable using `org.apache.beam.sdk.util.SerializableUtils#ensureSerializable(T)` or should we check using something else?

Comment (je-ik) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SerializableComparatorTest.java:36]: Yes, that is exactly the method to use. :+1:

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java
Comment (je-ik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java:41]: Makes sense. :+1:

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java
Comment (je-ik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SerializableComparator.java:42]: :+1:

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SerializableComparatorTest.java
Comment (kirupha2000) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SerializableComparatorTest.java]: Re-added the original tests and used `org.apache.beam.sdk.util.SerializableUtils#ensureSerializable(T)`.

**Review by je-ik — APPROVED** [review]
LGTM, I will merge after tests pass. Thanks for the contribution!

**Pull request merged**
